File: apps/api/plane/app/serializers/issue_type.py
# Copyright (c) 2023-present Plane Software, Inc. and contributors
# SPDX-License-Identifier: AGPL-3.0-only
# See the LICENSE file for details.

# Module imports
from .base import BaseSerializer
from plane.db.models import IssueType, ProjectIssueType


class IssueTypeSerializer(BaseSerializer):
    class Meta:
        model = IssueType
        fields = [
            "id",
            "name",
            "description",
            "logo_props",
            "is_epic",
            "is_default",
            "is_active",
            "level",
            "workspace",
            "external_source",
            "external_id",
            "created_at",
            "updated_at",
            "created_by",
            "updated_by",
        ]
        # `workspace` is derived from the URL/project and `is_epic` is managed
        # through the dedicated epic flow, so both are read-only here.
        read_only_fields = ["workspace", "is_epic"]


class IssueTypeLiteSerializer(BaseSerializer):
    class Meta:
        model = IssueType
        fields = ["id", "name", "logo_props", "is_epic", "is_default"]
        read_only_fields = fields


class ProjectIssueTypeSerializer(BaseSerializer):
    class Meta:
        model = ProjectIssueType
        fields = [
            "id",
            "issue_type",
            "project_id",
            "workspace_id",
            "level",
            "is_default",
        ]
        read_only_fields = ["workspace", "project"]

File: packages/types/src/issues/issue-type.ts
/**
 * Copyright (c) 2023-present Plane Software, Inc. and contributors
 * SPDX-License-Identifier: AGPL-3.0-only
 * See the LICENSE file for details.
 */

import type { TLogoProps } from "../common";

/**
 * A Work Item Type (a.k.a. issue type). Workspace-scoped, associated to projects
 * through {@link TProjectIssueType}. When `is_epic` is true the type represents an Epic.
 */
export type TIssueType = {
  id: string;
  name: string;
  description: string;
  logo_props: TLogoProps | undefined;
  is_epic: boolean;
  is_default: boolean;
  is_active: boolean;
  level: number;
  workspace: string;
  external_source: string | null;
  external_id: string | null;
  created_at: string;
  updated_at: string;
  created_by: string | null;
  updated_by: string | null;
};

/**
 * The association between a {@link TIssueType} and a project.
 */
export type TProjectIssueType = {
  id: string;
  issue_type: string;
  project_id: string;
  workspace_id: string;
  level: number;
  is_default: boolean;
};
